src/radical/pilot/raptor/master.py

```diff
@@ -427,105 +440,201 @@ def join(self):
+
+            self._req_put.put(tasks)
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _submit_executable_tasks(self, tasks):
+        '''
+        Submit tasks per given task description to the agent this
+        master is running in.
+        '''
+
+        if not tasks:
+            return
 
-        requests = list()
         for task in tasks:
 
-            # FIXME: abuse of arguments
-            req = json.loads(task['description']['arguments'][0])
+            td = task['description']
+
+            sbox = '%s/%s' % (os.environ['RP_TASK_SANDBOX'], td['uid'])
 
-            req['is_task'] = True
-            req['uid']     = task['uid']
-            req['task']    = task  # this duplicates the request :-/
-            requests.append(req)
+          # task['uid']               = td.get('uid')
+            task['state']             = rps.AGENT_STAGING_INPUT_PENDING
+            task['task_sandbox_path'] = sbox
+            task['task_sandbox']      = 'file://localhost/' + sbox
+            task['pilot_sandbox']     = os.environ['RP_PILOT_SANDBOX']
+            task['session_sandbox']   = os.environ['RP_SESSION_SANDBOX']
+            task['resource_sandbox']  = os.environ['RP_RESOURCE_SANDBOX']
+            task['pilot']             = os.environ['RP_PILOT_ID']
+            task['resources']         = {'cpu': td.get('cpu_processes', 1) *
+                                                td.get('cpu_threads',   1),
+                                         'gpu': td.get('gpu_processes', 0)}
+
+            # NOTE: the order of insert / state update relies on that order
+            #       being maintained through the component's message push,
+            #       the update worker's message receive up to the insertion
+            #       order into the update worker's DB bulk op.
+            self._log.debug('insert %s', td['uid'])
+            self.publish(rpc.STATE_PUBSUB, {'cmd': 'insert', 'arg': task})
+
+        self.advance(tasks, state=rps.AGENT_STAGING_INPUT_PENDING,
+                            publish=True, push=True)
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _request_cb(self, tasks):
+
+        tasks = ru.as_list(tasks)
+
+        self._log.debug('request_cb: %d', len(tasks))
 
         try:
-            filtered = self.request_cb(requests)
+            filtered = self.request_cb(tasks)
             if filtered:
-                self.request(filtered)
+                for task in filtered:
+                    self._log.debug('REQ cb: %s' % task['uid'])
+                self.submit_tasks(filtered)
 
         except:
             self._log.exception('request cb failed')
-            # FIXME: fail the request
+            # FIXME: fail tasks
 
 
+    # --------------------------------------------------------------------------
+    #
+    def request_cb(self, tasks):
+
+        # FIXME: document task format
```

Comment:
Will this be resolved for 1.14 or later?
Specifically, are `tasks` dictionaries or Task instances?

Reply:
Dictionaries in this case.  Better  documentation and typing will arrive in 1.15